src/common/contexts/provider.tsx

```diff
@@ -49,44 +74,118 @@ declare global {
+
+  const defaultProvider = isSupportedNetwork(defaultChainId) ? createProvider(defaultChainId) : undefined;
+  const [providerOrSigner, setProviderOrSigner] = useState<providers.Provider | Signer | undefined>(defaultProvider);
+
+  const updateProviderOrSigner = async (newProviderOrSigner: typeof providerOrSigner) => {
+    try {
+      if (!Signer.isSigner(newProviderOrSigner)) {
+        setIsConnected(false);
+      } else {
+        await (newProviderOrSigner as Signer).getAddress();
+        setIsConnected(true);
+      }
+    } catch (e) {
+      setIsConnected(false);
+    }
+    setProviderOrSigner(newProviderOrSigner);
+  };
 
+  const changeNetwork = async (chainId: ChainId) => {
+    if (!isSupportedNetwork(chainId)) throw new UnsupportedNetworkError(chainId);
+
+    const chainInfo = getChainInfo(chainId);
+
+    try {
+      const web3provider = getWeb3Provider();
+      await requestSwitchChain(chainId);
+      await updateProviderOrSigner(web3provider.getSigner());
+    } catch (e: unknown) {
+      if (e instanceof NoMetaMaskError) {
+        console.warn(e.message);
+        await updateProviderOrSigner(createProvider(chainInfo.chainId));
+      } else {
+        throw e;
+      }
+    }
+  };
+
+  const getProvider = useCallback(() => {
+    if (providers.Provider.isProvider(providerOrSigner)) return providerOrSigner;
+    if (Signer.isSigner(providerOrSigner)) return providerOrSigner.provider;
+    return undefined;
+  }, [providerOrSigner]);
+
+  const getSigner = useCallback(
+    () => (isConnected ? (providerOrSigner as Signer) : undefined),
+    [isConnected, providerOrSigner]
+  );
+
+  const getTransactor = useCallback(() => getSigner() ?? getProvider(), [getProvider, getSigner]);
+
+  const getWeb3Provider = () => {
+    const { ethereum, web3 } = window;
     const metamaskExtensionNotFound = typeof ethereum === "undefined" || typeof web3 === "undefined";
-    if (metamaskExtensionNotFound) throw new Error(MessageTitle.NO_METAMASK);
+    if (metamaskExtensionNotFound || !ethereum.request) throw new NoMetaMaskError();
 
-    await ethereum.enable();
     const injectedWeb3 = ethereum || web3.currentProvider;
     if (!injectedWeb3) throw new Error("No injected web3 provider found");
-    const web3provider = new ethers.providers.Web3Provider(injectedWeb3);
-    const signer = web3provider.getSigner();
-    const web3account = (await web3provider.listAccounts())[0];
+    return new ethers.providers.Web3Provider(injectedWeb3, "any");
+  };
+
+  const requestSwitchChain = async (chainId: ChainId) => {
+    const { ethereum } = window;
+    if (!ethereum || !ethereum.request) return;
+    try {
+      await ethereum.request({
+        method: "wallet_switchEthereumChain",
+        params: [{ chainId: `0x${chainId.toString(16)}` }],
+      });
+    } catch (e: any) {
+      if (e.code === -32601) {
+        // Possibly on localhost which doesn't support the call
```

Comment:
um just curious, where to see the list of potential error codes?

Reply:
Yea, have. They are here: https://github.com/MetaMask/eth-rpc-errors/blob/main/src/error-constants.ts#L11